src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -552,41 +556,141 @@ def calculate_dense(self):
+            measurements = [
+                gates.M(factor.target_qubit, basis=factor.gate.__class__)
+                for factor in set(term.factors)
+            ]
+            circ_copy = circuit.copy(True)
+            [circ_copy.add(m) for m in measurements]
+            rotated_circuits.append(circ_copy)
+            # map the original qubits into 0, 1, 2, ...
+            q_maps.append(
+                dict(
+                    zip(
+                        [qubit_map[q] for q in term.target_qubits],
+                        range(len(term.target_qubits)),
+                    )
+                )
+            )
+        frequencies = [
+            result.frequencies()
+            for result in self.backend.execute_circuits(rotated_circuits, nshots=nshots)
+        ]
+        return sum(
+            [
+                coeff * obs._exp_from_freq(freq, qmap)
+                for coeff, freq, obs, qmap in zip(
+                    coefficients, frequencies, Z_observables, q_maps
+                )
+            ]
+        )
+
+    def _exp_from_freq(self, freq: dict, qubit_map: dict) -> float:
+        """
+        Calculate the expectation value from some frequencies.
+        The observable has to be diagonal in the computational basis.
+
+        Args:
+            freq (dict): input frequencies.
+            qubit_map (dict): qubit map.
+
+        Returns:
+            (float): the calculated expectation value.
+        """
+        for term in self.terms:
             # pylint: disable=E1101
             for factor in term.factors:
                 if not isinstance(factor, Z):
                     raise_error(
                         NotImplementedError, "Observable is not a Z Pauli string."
                     )
-            if len(term.factors) != len(set(term.factors)):
-                raise_error(NotImplementedError, "Z^k is not implemented since Z^2=I.")
+
         keys = list(freq.keys())
         counts = self.backend.cast(list(freq.values()), self.backend.precision) / sum(
             freq.values()
         )
-        qubits = []
-        for term in terms:
-            qubits_term = []
-            for k in term.target_qubits:
-                qubits_term.append(k)
-            qubits.append(qubits_term)
+        expvals = []
+        for term in self.terms:
+            qubits = term.target_qubits
+            expvals.extend(
+                [
+                    term.coefficient.real
+                    * (-1) ** [state[qubit_map[q]] for q in qubits].count("1")
+                    for state in keys
+                ]
+            )
+        expvals = self.backend.cast(expvals, dtype=counts.dtype).reshape(
+            len(self.terms), len(freq)
+        )
+        return self.backend.np.sum(expvals @ counts.T) + self.constant.real
+
+    def expectation_from_samples(
+        self, data: Union[dict, "Circuit"], qubit_map: dict = None, nshots: int = None
```

Comment:
This is also up for discussion, however it may even be cleaner to expose two separate functions: `expectation_from_samples` and `expectation_from_circuit`, instead of accepting two different types.